File: cpp/data_structures/linear/deque/deque.cpp
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

struct CircularDeque {
    vector<int> buf;
    int cap, front, sz;
    explicit CircularDeque(int c = 8) : cap(max(4, c)), front(0), sz(0) { buf.assign(cap, 0); }

    void grow() {
        vector<int> nb(cap * 2, 0);
        for (int i = 0; i < sz; ++i) nb[i] = buf[(front + i) % cap];
        buf.swap(nb);
        cap = (int)buf.size();
        front = 0;
    }

    void push_back(int x) {
        if (sz == cap) grow();
        int idx = (front + sz) % cap;
        buf[idx] = x;
        ++sz;
    }

    void push_front(int x) {
        if (sz == cap) grow();
        front = (front - 1 + cap) % cap;
        buf[front] = x;
        ++sz;
    }

    int pop_back() {
        if (sz == 0) throw underflow_error("pop_back");
        int idx = (front + sz - 1) % cap;
        --sz;
        return buf[idx];
    }

    int pop_front() {
        if (sz == 0) throw underflow_error("pop_front");
        int v = buf[front];
        front = (front + 1) % cap;
        --sz;
        return v;
    }
};

int main() {
    CircularDeque d(4);
    d.push_back(1);
    d.push_front(0);
    d.push_back(2);
    assert(d.pop_front() == 0);
    assert(d.pop_back() == 2);
    assert(d.pop_front() == 1);
    cout << "Deque OK" << endl;
    return 0;
}

File: python/data_structures/linear/deque/deque.py
"""循环数组双端队列。"""

from __future__ import annotations


class CircularDeque:
    def __init__(self, cap: int = 8) -> None:
        if cap < 4:
            cap = 4
        self._cap = cap
        self._buf: list[object | None] = [None] * cap
        self._front = 0
        self._size = 0

    def __len__(self) -> int:
        return self._size

    def push_back(self, x: object) -> None:
        if self._size == self._cap:
            self._grow()
        idx = (self._front + self._size) % self._cap
        self._buf[idx] = x
        self._size += 1

    def push_front(self, x: object) -> None:
        if self._size == self._cap:
            self._grow()
        self._front = (self._front - 1 + self._cap) % self._cap
        self._buf[self._front] = x
        self._size += 1

    def pop_back(self) -> object:
        if self._size == 0:
            raise IndexError("pop_back empty")
        idx = (self._front + self._size - 1) % self._cap
        v = self._buf[idx]
        self._buf[idx] = None
        self._size -= 1
        return v  # type: ignore[return-value]

    def pop_front(self) -> object:
        if self._size == 0:
            raise IndexError("pop_front empty")
        v = self._buf[self._front]
        self._buf[self._front] = None
        self._front = (self._front + 1) % self._cap
        self._size -= 1
        return v  # type: ignore[return-value]

    def _grow(self) -> None:
        new_cap = self._cap * 2
        new_buf: list[object | None] = [None] * new_cap
        for i in range(self._size):
            new_buf[i] = self._buf[(self._front + i) % self._cap]
        self._buf = new_buf
        self._cap = new_cap
        self._front = 0


if __name__ == "__main__":
    d = CircularDeque(4)
    d.push_back(1)
    d.push_front(0)
    d.push_back(2)
    assert d.pop_front() == 0
    assert d.pop_back() == 2
    assert d.pop_front() == 1
    d2 = CircularDeque(4)
    try:
        d2.pop_front()
        raise AssertionError("expected IndexError")
    except IndexError:
        pass
    print("Deque OK")
